python/pyarrow/_dataset.pyx

```diff
@@ -0,0 +1,1381 @@
+            filesystem.unwrap(),
+            selector.unwrap(),
+            format.unwrap(),
+            options.unwrap()
+        )
+        self.init(GetResultValue(result))
+
+    cdef init(self, shared_ptr[CDataSourceDiscovery]& sp):
+        DataSourceDiscovery.init(self, sp)
+        self.filesystem_discovery = <CFileSystemDataSourceDiscovery*> sp.get()
+
+
+cdef class FileSource:
+
+    cdef:
+        shared_ptr[CFileSource] wrapped
+        CFileSource* source
+
+    cdef readonly:
+        FileSystem fs
+
+    def __init__(self, path, FileSystem filesystem not None, compression=None):
+        cdef shared_ptr[CFileSource] source
+
+        # need to hold a reference for the filesystem
+        self.fs = filesystem
+
+        source.reset(new CFileSource(
+            tobytes(path),
+            self.fs.unwrap().get(),
+            _get_compression_type(compression)
+        ))
+
+        self.init(source)
+
+    cdef init(self, shared_ptr[CFileSource]& sp):
+        self.wrapped = sp
+        self.source = sp.get()
+
+    @staticmethod
+    cdef wrap(shared_ptr[CFileSource]& sp):
+        cdef FileSource self = FileSource.__new__(FileSource)
+        self.init(sp)
+        return self
+
+    cdef inline shared_ptr[CFileSource] unwrap(self):
+        return self.wrapped
+
+    def equals(self, FileSource other):
+        return deref(self.source) == deref(other.unwrap())
+
+    def __eq__(self, other):
+        try:
+            return self.equals(other)
+        except TypeError:
+            return NotImplemented
+
+    @property
+    def compression(self):
+        return self.wrapped.get().compression()
+
+    @property
+    def path(self):
+        return frombytes(self.wrapped.get().path())
+
+    # def open(self):
+    #     CStatus Open(shared_ptr[CRandomAccessFile]* out)
+
+
+cdef class DataFragment:
+    """A granular piece of a Dataset such as an individual file.
+
+    It can be read/scanned separately from other fragments.
+    A DataFragment yields a collection of RecordBatch, encapsulated in one or
+    more ScanTasks.
+    """
+    cdef:
+        shared_ptr[CDataFragment] wrapped
+        CDataFragment* fragment
+
+    def __init__(self):
+        _forbid_instantiation(self.__class__)
+
+    cdef void init(self, const shared_ptr[CDataFragment]& sp):
+        self.wrapped = sp
+        self.fragment = sp.get()
+
+    @staticmethod
+    cdef wrap(shared_ptr[CDataFragment]& sp):
+        cdef DataFragment self
+
+        typ = frombytes(sp.get().type())
+        if typ == 'simple_data_fragment':
+            self = SimpleDataFragment.__new__(SimpleDataFragment)
+        elif typ == 'parquet_data_fragment':
+            self = ParquetDataFragment.__new__(ParquetDataFragment)
+        else:
+            raise TypeError(typ)
+
+        self.init(sp)
+        return self
+
+    cdef inline shared_ptr[CDataFragment] unwrap(self):
+        return self.wrapped
+
+    def scan(self, ScanContext context=None):
```

Comment:
And let the `scan()` here directly iterate thought the scan tasks, execute them and returns RecordBatches? 

(that would certainly make the python api easier to use for the basic use case if you can go directly from Scanner -> list of RecordBatches)

Reply:
The scheduling will only happen internally? Don't we want to let the user to have custom schedulers? Besides that `scanner.scan()` produces ScanTasks, although we can directly produces recordbatches too.

Reply:
Then it would be nice from the C++ API to provide such a method. IMO I'd be nice to expose the scan tasks to let user defined scheduling (in the same process) and a function for convenience which directly yields the record batches.